lahja/asyncio/endpoint.py

```diff
@@ -242,13 +252,12 @@ def can_send_item(self, item: BaseEvent, config: Optional[BroadcastConfig]) -> b
     async def send_message(self, message: Msg) -> None:
         await self.conn.send_message(message)
 
-    async def wait_until_subscription_received(self) -> None:
-        async with self._received_subscription:
-            await self._received_subscription.wait()
-
     async def wait_until_subscribed_to(self, event: Type[BaseEvent]) -> None:
-        while event not in self.subscribed_messages:
-            await self.wait_until_subscription_received()
+        async with self._received_subscription:
+            while True:
+                if event in self.subscribed_messages:
+                    return
+                await self._received_subscription.wait()
 
 
 @asynccontextmanager  # type: ignore
```

Comment:
As of this change `_handle_client` and `_handle_server` are the same method and together they constitute the only callers of `run_remote_endpoint`. Maybe they could all be merged into one method (untested):

```python
async def run_remote_endpoint(remote: RemoteEndpoint) -> None
    await remote.start()
    try:
        await remote.wait_stopped()
    finally:
        await remote.stop()
```

Reply:
good catch.